// @ironcorelabs/recrypt-wasm-binding/recrypt_wasm_binding_bg.js
import * as wasm from './recrypt_wasm_binding_bg.wasm';

const heap = new Array(32).fill(undefined);

heap.push(undefined, null, true, false);

function getObject(idx) { return heap[idx]; }

let heap_next = heap.length;

function dropObject(idx) {
    if (idx < 36) return;
    heap[idx] = heap_next;
    heap_next = idx;
}

function takeObject(idx) {
    const ret = getObject(idx);
    dropObject(idx);
    return ret;
}

function addHeapObject(obj) {
    if (heap_next === heap.length) heap.push(heap.length + 1);
    const idx = heap_next;
    heap_next = heap[idx];

    heap[idx] = obj;
    return idx;
}

const lTextDecoder = typeof TextDecoder === 'undefined' ? (0, module.require)('util').TextDecoder : TextDecoder;

let cachedTextDecoder = new lTextDecoder('utf-8', { ignoreBOM: true, fatal: true });

cachedTextDecoder.decode();

let cachegetUint8Memory0 = null;
function getUint8Memory0() {
    if (cachegetUint8Memory0 === null || cachegetUint8Memory0.buffer !== wasm.memory.buffer) {
        cachegetUint8Memory0 = new Uint8Array(wasm.memory.buffer);
    }
    return cachegetUint8Memory0;
}

function getStringFromWasm0(ptr, len) {
    return cachedTextDecoder.decode(getUint8Memory0().subarray(ptr, ptr + len));
}

let WASM_VECTOR_LEN = 0;

const lTextEncoder = typeof TextEncoder === 'undefined' ? (0, module.require)('util').TextEncoder : TextEncoder;

let cachedTextEncoder = new lTextEncoder('utf-8');

const encodeString = (typeof cachedTextEncoder.encodeInto === 'function'
    ? function (arg, view) {
    return cachedTextEncoder.encodeInto(arg, view);
}
    : function (arg, view) {
    const buf = cachedTextEncoder.encode(arg);
    view.set(buf);
    return {
        read: arg.length,
        written: buf.length
    };
});

function passStringToWasm0(arg, malloc, realloc) {

    if (realloc === undefined) {
        const buf = cachedTextEncoder.encode(arg);
        const ptr = malloc(buf.length);
        getUint8Memory0().subarray(ptr, ptr + buf.length).set(buf);
        WASM_VECTOR_LEN = buf.length;
        return ptr;
    }

    let len = arg.length;
    let ptr = malloc(len);

    const mem = getUint8Memory0();

    let offset = 0;

    for (; offset < len; offset++) {
        const code = arg.charCodeAt(offset);
        if (code > 0x7F) break;
        mem[ptr + offset] = code;
    }

    if (offset !== len) {
        if (offset !== 0) {
            arg = arg.slice(offset);
        }
        ptr = realloc(ptr, len, len = offset + arg.length * 3);
        const view = getUint8Memory0().subarray(ptr + offset, ptr + len);
        const ret = encodeString(arg, view);

        offset += ret.written;
    }

    WASM_VECTOR_LEN = offset;
    return ptr;
}

let cachegetInt32Memory0 = null;
function getInt32Memory0() {
    if (cachegetInt32Memory0 === null || cachegetInt32Memory0.buffer !== wasm.memory.buffer) {
        cachegetInt32Memory0 = new Int32Array(wasm.memory.buffer);
    }
    return cachegetInt32Memory0;
}

function handleError(f, args) {
    try {
        return f.apply(this, args);
    } catch (e) {
        wasm.__wbindgen_exn_store(addHeapObject(e));
    }
}

function getArrayU8FromWasm0(ptr, len) {
    return getUint8Memory0().subarray(ptr / 1, ptr / 1 + len);
}

function passArray8ToWasm0(arg, malloc) {
    const ptr = malloc(arg.length * 1);
    getUint8Memory0().set(arg, ptr / 1);
    WASM_VECTOR_LEN = arg.length;
    return ptr;
}

let stack_pointer = 32;

function addBorrowedObject(obj) {
    if (stack_pointer == 1) throw new Error('out of js stack');
    heap[--stack_pointer] = obj;
    return stack_pointer;
}

function isLikeNone(x) {
    return x === undefined || x === null;
}

let cachegetUint32Memory0 = null;
function getUint32Memory0() {
    if (cachegetUint32Memory0 === null || cachegetUint32Memory0.buffer !== wasm.memory.buffer) {
        cachegetUint32Memory0 = new Uint32Array(wasm.memory.buffer);
    }
    return cachegetUint32Memory0;
}

function getArrayU32FromWasm0(ptr, len) {
    return getUint32Memory0().subarray(ptr / 4, ptr / 4 + len);
}
/**
*
* * Hash the provided transform key into a buffer of bytes. The various transform key object fields are concatenated
* * in a specific order in order for transform keys to be signed over.
*
* @param {any} transform_key_obj
* @returns {Uint8Array}
*/
export function transformKeyToBytes256(transform_key_obj) {
    try {
        const retptr = wasm.__wbindgen_add_to_stack_pointer(-16);
        wasm.transformKeyToBytes256(retptr, addBorrowedObject(transform_key_obj));
        var r0 = getInt32Memory0()[retptr / 4 + 0];
        var r1 = getInt32Memory0()[retptr / 4 + 1];
        var v0 = getArrayU8FromWasm0(r0, r1).slice();
        wasm.__wbindgen_free(r0, r1 * 1);
        return v0;
    } finally {
        wasm.__wbindgen_add_to_stack_pointer(16);
        heap[stack_pointer++] = undefined;
    }
}

/**
*
* * Augment the provided transform key with the provided private key. Returns an augmented TransformKey object.
*
* @param {any} transform_key_obj
* @param {Uint8Array} private_key
* @returns {any}
*/
export function augmentTransformKey256(transform_key_obj, private_key) {
    try {
        var ptr0 = passArray8ToWasm0(private_key, wasm.__wbindgen_malloc);
        var len0 = WASM_VECTOR_LEN;
        var ret = wasm.augmentTransformKey256(addBorrowedObject(transform_key_obj), ptr0, len0);
        return takeObject(ret);
    } finally {
        heap[stack_pointer++] = undefined;
    }
}

/**
*
* * Augment the provided public key with the other provided public key. Returns a new augmented PublicKey object.
*
* @param {any} current_public_key_obj
* @param {any} other_public_key_obj
* @returns {any}
*/
export function augmentPublicKey256(current_public_key_obj, other_public_key_obj) {
    try {
        var ret = wasm.augmentPublicKey256(addBorrowedObject(current_public_key_obj), addBorrowedObject(other_public_key_obj));
        return takeObject(ret);
    } finally {
        heap[stack_pointer++] = undefined;
        heap[stack_pointer++] = undefined;
    }
}

/**
*
* * Add the two provided private keys together and return the bytes of a new PrivateKey.
*
* @param {Uint8Array} private_key_a
* @param {Uint8Array} private_key_b
* @returns {Uint8Array}
*/
export function addPrivateKeys(private_key_a, private_key_b) {
    try {
        const retptr = wasm.__wbindgen_add_to_stack_pointer(-16);
        var ptr0 = passArray8ToWasm0(private_key_a, wasm.__wbindgen_malloc);
        var len0 = WASM_VECTOR_LEN;
        var ptr1 = passArray8ToWasm0(private_key_b, wasm.__wbindgen_malloc);
        var len1 = WASM_VECTOR_LEN;
        wasm.addPrivateKeys(retptr, ptr0, len0, ptr1, len1);
        var r0 = getInt32Memory0()[retptr / 4 + 0];
        var r1 = getInt32Memory0()[retptr / 4 + 1];
        var v2 = getArrayU8FromWasm0(r0, r1).slice();
        wasm.__wbindgen_free(r0, r1 * 1);
        return v2;
    } finally {
        wasm.__wbindgen_add_to_stack_pointer(16);
    }
}

/**
*
* * Subtract the first provided private key from the second provided private key. Returns the bytes of a new PrivateKey.
*
* @param {Uint8Array} private_key_a
* @param {Uint8Array} private_key_b
* @returns {Uint8Array}
*/
export function subtractPrivateKeys(private_key_a, private_key_b) {
    try {
        const retptr = wasm.__wbindgen_add_to_stack_pointer(-16);
        var ptr0 = passArray8ToWasm0(private_key_a, wasm.__wbindgen_malloc);
        var len0 = WASM_VECTOR_LEN;
        var ptr1 = passArray8ToWasm0(private_key_b, wasm.__wbindgen_malloc);
        var len1 = WASM_VECTOR_LEN;
        wasm.subtractPrivateKeys(retptr, ptr0, len0, ptr1, len1);
        var r0 = getInt32Memory0()[retptr / 4 + 0];
        var r1 = getInt32Memory0()[retptr / 4 + 1];
        var v2 = getArrayU8FromWasm0(r0, r1).slice();
        wasm.__wbindgen_free(r0, r1 * 1);
        return v2;
    } finally {
        wasm.__wbindgen_add_to_stack_pointer(16);
    }
}

/**
*/
export class Api256 {

    static __wrap(ptr) {
        const obj = Object.create(Api256.prototype);
        obj.ptr = ptr;

        return obj;
    }

    __destroy_into_raw() {
        const ptr = this.ptr;
        this.ptr = 0;

        return ptr;
    }

    free() {
        const ptr = this.__destroy_into_raw();
        wasm.__wbg_api256_free(ptr);
    }
    /**
    */
    constructor() {
        var ret = wasm.api256_new();
        return Api256.__wrap(ret);
    }
    /**
    *
    *     * Generate a new Recrypt key pair with both a public and private key and return both
    *     * as Uint8Arrays.
    *
    * @returns {any}
    */
    generateKeyPair() {
        var ret = wasm.api256_generateKeyPair(this.ptr);
        return takeObject(ret);
    }
    /**
    *
    *     * Generate and return an ed25519 signing public and private key.
    *
    * @returns {any}
    */
    generateEd25519KeyPair() {
        var ret = wasm.api256_generateEd25519KeyPair(this.ptr);
        return takeObject(ret);
    }
    /**
    *
    *     * Sign the provided variable length message with the provided signing private key. Returns a 64 byte signature.
    *
    * @param {Uint8Array} signing_private_key
    * @param {Uint8Array} message
    * @returns {Uint8Array}
    */
    ed25519Sign(signing_private_key, message) {
        try {
            const retptr = wasm.__wbindgen_add_to_stack_pointer(-16);
            var ptr0 = passArray8ToWasm0(signing_private_key, wasm.__wbindgen_malloc);
            var len0 = WASM_VECTOR_LEN;
            var ptr1 = passArray8ToWasm0(message, wasm.__wbindgen_malloc);
            var len1 = WASM_VECTOR_LEN;
            wasm.api256_ed25519Sign(retptr, this.ptr, ptr0, len0, ptr1, len1);
            var r0 = getInt32Memory0()[retptr / 4 + 0];
            var r1 = getInt32Memory0()[retptr / 4 + 1];
            var v2 = getArrayU8FromWasm0(r0, r1).slice();
            wasm.__wbindgen_free(r0, r1 * 1);
            return v2;
        } finally {
            wasm.__wbindgen_add_to_stack_pointer(16);
        }
    }
    /**
    *
    *     * Verify the provided signature given the signing public key and the message to verify. Returns a boolean of whether the message was verified.
    *
    * @param {Uint8Array} signing_public_key
    * @param {Uint8Array} message
    * @param {Uint8Array} signature
    * @returns {boolean}
    */
    ed25519Verify(signing_public_key, message, signature) {
        var ptr0 = passArray8ToWasm0(signing_public_key, wasm.__wbindgen_malloc);
        var len0 = WASM_VECTOR_LEN;
        var ptr1 = passArray8ToWasm0(message, wasm.__wbindgen_malloc);
        var len1 = WASM_VECTOR_LEN;
        var ptr2 = passArray8ToWasm0(signature, wasm.__wbindgen_malloc);
        var len2 = WASM_VECTOR_LEN;
        var ret = wasm.api256_ed25519Verify(this.ptr, ptr0, len0, ptr1, len1, ptr2, len2);
        return ret !== 0;
    }
    /**
    *
    *     * Compute an ed25519 signing public key given the signing private key.
    *
    * @param {Uint8Array} signing_private_key
    * @returns {Uint8Array}
    */
    computeEd25519PublicKey(signing_private_key) {
        try {
            const retptr = wasm.__wbindgen_add_to_stack_pointer(-16);
            var ptr0 = passArray8ToWasm0(signing_private_key, wasm.__wbindgen_malloc);
            var len0 = WASM_VECTOR_LEN;
            wasm.api256_computeEd25519PublicKey(retptr, this.ptr, ptr0, len0);
            var r0 = getInt32Memory0()[retptr / 4 + 0];
            var r1 = getInt32Memory0()[retptr / 4 + 1];
            var v1 = getArrayU8FromWasm0(r0, r1).slice();
            wasm.__wbindgen_free(r0, r1 * 1);
            return v1;
        } finally {
            wasm.__wbindgen_add_to_stack_pointer(16);
        }
    }
    /**
    *
    *     * Generate a new Recrypt plaintext and return it as a Uint8Array.
    *
    * @returns {Uint8Array}
    */
    generatePlaintext() {
        try {
            const retptr = wasm.__wbindgen_add_to_stack_pointer(-16);
            wasm.api256_generatePlaintext(retptr, this.ptr);
            var r0 = getInt32Memory0()[retptr / 4 + 0];
            var r1 = getInt32Memory0()[retptr / 4 + 1];
            var v0 = getArrayU8FromWasm0(r0, r1).slice();
            wasm.__wbindgen_free(r0, r1 * 1);
            return v0;
        } finally {
            wasm.__wbindgen_add_to_stack_pointer(16);
        }
    }
    /**
    *
    *     * Generate a transform key from the provided private key to the provided public key and use the provided signing key pair. Returns
    *     * a JsObject which contains all the necessary keys for the transform key as Uint8Arrays.
    *
    * @param {Uint8Array} from_private_key
    * @param {any} to_public_key
    * @param {Uint8Array} private_signing_key
    * @returns {any}
    */
    generateTransformKey(from_private_key, to_public_key, private_signing_key) {
        try {
            var ptr0 = passArray8ToWasm0(from_private_key, wasm.__wbindgen_malloc);
            var len0 = WASM_VECTOR_LEN;
            var ptr1 = passArray8ToWasm0(private_signing_key, wasm.__wbindgen_malloc);
            var len1 = WASM_VECTOR_LEN;
            var ret = wasm.api256_generateTransformKey(this.ptr, ptr0, len0, addBorrowedObject(to_public_key), ptr1, len1);
            return takeObject(ret);
        } finally {
            heap[stack_pointer++] = undefined;
        }
    }
    /**
    *
    *     * Compute the private key for the provided public key. Returns a public key object in x/y form.
    *
    * @param {Uint8Array} private_key
    * @returns {any}
    */
    computePublicKey(private_key) {
        var ptr0 = passArray8ToWasm0(private_key, wasm.__wbindgen_malloc);
        var len0 = WASM_VECTOR_LEN;
        var ret = wasm.api256_computePublicKey(this.ptr, ptr0, len0);
        return takeObject(ret);
    }
    /**
    *
    *     * Hash the incoming bytes to _exactly_ 32 bytes. Useful for generating PrivateKeys, among other things.
    *
    * @param {Uint8Array} hashable_bytes
    * @returns {Uint8Array}
    */
    hash256(hashable_bytes) {
        try {
            const retptr = wasm.__wbindgen_add_to_stack_pointer(-16);
            var ptr0 = passArray8ToWasm0(hashable_bytes, wasm.__wbindgen_malloc);
            var len0 = WASM_VECTOR_LEN;
            wasm.api256_hash256(retptr, this.ptr, ptr0, len0);
            var r0 = getInt32Memory0()[retptr / 4 + 0];
            var r1 = getInt32Memory0()[retptr / 4 + 1];
            var v1 = getArrayU8FromWasm0(r0, r1).slice();
            wasm.__wbindgen_free(r0, r1 * 1);
            return v1;
        } finally {
            wasm.__wbindgen_add_to_stack_pointer(16);
        }
    }
    /**
    *
    *     * Derives a symmetric key from the provided plaintext.
    *
    * @param {Uint8Array} plaintext
    * @returns {Uint8Array}
    */
    deriveSymmetricKey(plaintext) {
        try {
            const retptr = wasm.__wbindgen_add_to_stack_pointer(-16);
            var ptr0 = passArray8ToWasm0(plaintext, wasm.__wbindgen_malloc);
            var len0 = WASM_VECTOR_LEN;
            wasm.api256_deriveSymmetricKey(retptr, this.ptr, ptr0, len0);
            var r0 = getInt32Memory0()[retptr / 4 + 0];
            var r1 = getInt32Memory0()[retptr / 4 + 1];
            var v1 = getArrayU8FromWasm0(r0, r1).slice();
            wasm.__wbindgen_free(r0, r1 * 1);
            return v1;
        } finally {
            wasm.__wbindgen_add_to_stack_pointer(16);
        }
    }
    /**
    *
    *     * Encrypt the provided plaintext to the provided public key. Use the provided signing key pair to sign the encrypted value. Returns
    *     * a complex object which can be passed into the decrypt function to get the provided plaintext back.
    *
    * @param {Uint8Array} plaintext
    * @param {any} to_public_key
    * @param {Uint8Array} private_signing_key
    * @returns {any}
    */
    encrypt(plaintext, to_public_key, private_signing_key) {
        try {
            var ptr0 = passArray8ToWasm0(plaintext, wasm.__wbindgen_malloc);
            var len0 = WASM_VECTOR_LEN;
            var ptr1 = passArray8ToWasm0(private_signing_key, wasm.__wbindgen_malloc);
            var len1 = WASM_VECTOR_LEN;
            var ret = wasm.api256_encrypt(this.ptr, ptr0, len0, addBorrowedObject(to_public_key), ptr1, len1);
            return takeObject(ret);
        } finally {
            heap[stack_pointer++] = undefined;
        }
    }
    /**
    *
    *     * Transform the provided EncryptedValue JS object using the provided TransformKey JS object. Returns a transformed
    *     * EncryptedValue. Use the provided signing key pair to sign the encrypted value.
    *
    * @param {any} encrypted_value
    * @param {any} transform_key
    * @param {Uint8Array} private_signing_key
    * @returns {any}
    */
    transform(encrypted_value, transform_key, private_signing_key) {
        try {
            var ptr0 = passArray8ToWasm0(private_signing_key, wasm.__wbindgen_malloc);
            var len0 = WASM_VECTOR_LEN;
            var ret = wasm.api256_transform(this.ptr, addBorrowedObject(encrypted_value), addBorrowedObject(transform_key), ptr0, len0);
            return takeObject(ret);
        } finally {
            heap[stack_pointer++] = undefined;
            heap[stack_pointer++] = undefined;
        }
    }
    /**
    *
    *     * Decrypt the provided encrypted value object using the provided private key. Returns the encrypted plaintext bytes as a Uint8Array.
    *
    * @param {any} encrypted_value
    * @param {Uint8Array} private_key
    * @returns {Uint8Array}
    */
    decrypt(encrypted_value, private_key) {
        try {
            const retptr = wasm.__wbindgen_add_to_stack_pointer(-16);
            var ptr0 = passArray8ToWasm0(private_key, wasm.__wbindgen_malloc);
            var len0 = WASM_VECTOR_LEN;
            wasm.api256_decrypt(retptr, this.ptr, addBorrowedObject(encrypted_value), ptr0, len0);
            var r0 = getInt32Memory0()[retptr / 4 + 0];
            var r1 = getInt32Memory0()[retptr / 4 + 1];
            var v1 = getArrayU8FromWasm0(r0, r1).slice();
            wasm.__wbindgen_free(r0, r1 * 1);
            return v1;
        } finally {
            wasm.__wbindgen_add_to_stack_pointer(16);
            heap[stack_pointer++] = undefined;
        }
    }
    /**
    *
    *     * Schnorr sign the provided message using the provided public and private key. Returns a 64-byte signature.
    *
    * @param {Uint8Array} private_key
    * @param {any} public_key_obj
    * @param {Uint8Array} message
    * @returns {Uint8Array}
    */
    schnorrSign(private_key, public_key_obj, message) {
        try {
            const retptr = wasm.__wbindgen_add_to_stack_pointer(-16);
            var ptr0 = passArray8ToWasm0(private_key, wasm.__wbindgen_malloc);
            var len0 = WASM_VECTOR_LEN;
            var ptr1 = passArray8ToWasm0(message, wasm.__wbindgen_malloc);
            var len1 = WASM_VECTOR_LEN;
            wasm.api256_schnorrSign(retptr, this.ptr, ptr0, len0, addBorrowedObject(public_key_obj), ptr1, len1);
            var r0 = getInt32Memory0()[retptr / 4 + 0];
            var r1 = getInt32Memory0()[retptr / 4 + 1];
            var v2 = getArrayU8FromWasm0(r0, r1).slice();
            wasm.__wbindgen_free(r0, r1 * 1);
            return v2;
        } finally {
            wasm.__wbindgen_add_to_stack_pointer(16);
            heap[stack_pointer++] = undefined;
        }
    }
    /**
    *
    *     * Verify the provided signature is valid by providing the public key, augmented private key, and original signed message content. Returns
    *     * a boolean that denotes if the signature was verified.
    *
    * @param {any} public_key_obj
    * @param {Uint8Array | undefined} augmented_private_key
    * @param {Uint8Array} message
    * @param {Uint8Array} signature
    * @returns {boolean}
    */
    schnorrVerify(public_key_obj, augmented_private_key, message, signature) {
        try {
            var ptr0 = isLikeNone(augmented_private_key) ? 0 : passArray8ToWasm0(augmented_private_key, wasm.__wbindgen_malloc);
            var len0 = WASM_VECTOR_LEN;
            var ptr1 = passArray8ToWasm0(message, wasm.__wbindgen_malloc);
            var len1 = WASM_VECTOR_LEN;
            var ptr2 = passArray8ToWasm0(signature, wasm.__wbindgen_malloc);
            var len2 = WASM_VECTOR_LEN;
            var ret = wasm.api256_schnorrVerify(this.ptr, addBorrowedObject(public_key_obj), ptr0, len0, ptr1, len1, ptr2, len2);
            return ret !== 0;
        } finally {
            heap[stack_pointer++] = undefined;
        }
    }
}
/**
*/
export class EncryptedSearch {

    static __wrap(ptr) {
        const obj = Object.create(EncryptedSearch.prototype);
        obj.ptr = ptr;

        return obj;
    }

    __destroy_into_raw() {
        const ptr = this.ptr;
        this.ptr = 0;

        return ptr;
    }

    free() {
        const ptr = this.__destroy_into_raw();
        wasm.__wbg_encryptedsearch_free(ptr);
    }
    /**
    */
    constructor() {
        var ret = wasm.encryptedsearch_new();
        return EncryptedSearch.__wrap(ret);
    }
    /**
    *
    *     * Hashes all possible tri-grams for the given string. The values will be prefixed with the partition_id and salt before being hashed.
    *
    * @param {string} s
    * @param {Uint8Array} salt
    * @param {string | undefined} partition_id
    * @returns {Uint32Array}
    */
    generateHashesForString(s, salt, partition_id) {
        try {
            const retptr = wasm.__wbindgen_add_to_stack_pointer(-16);
            var ptr0 = passStringToWasm0(s, wasm.__wbindgen_malloc, wasm.__wbindgen_realloc);
            var len0 = WASM_VECTOR_LEN;
            var ptr1 = passArray8ToWasm0(salt, wasm.__wbindgen_malloc);
            var len1 = WASM_VECTOR_LEN;
            var ptr2 = isLikeNone(partition_id) ? 0 : passStringToWasm0(partition_id, wasm.__wbindgen_malloc, wasm.__wbindgen_realloc);
            var len2 = WASM_VECTOR_LEN;
            wasm.encryptedsearch_generateHashesForString(retptr, this.ptr, ptr0, len0, ptr1, len1, ptr2, len2);
            var r0 = getInt32Memory0()[retptr / 4 + 0];
            var r1 = getInt32Memory0()[retptr / 4 + 1];
            var v3 = getArrayU32FromWasm0(r0, r1).slice();
            wasm.__wbindgen_free(r0, r1 * 4);
            return v3;
        } finally {
            wasm.__wbindgen_add_to_stack_pointer(16);
        }
    }
    /**
    *
    *     * Hashes all possible tri-grams for the given string. The values will be prefixed with the partition_id and salt before
    *     * being hashed. This function will also add some random entries to the result to not expose how many tri-grams were actually found.
    *
    * @param {string} s
    * @param {Uint8Array} salt
    * @param {string | undefined} partition_id
    * @returns {Uint32Array}
    */
    generateHashesForStringWithPadding(s, salt, partition_id) {
        try {
            const retptr = wasm.__wbindgen_add_to_stack_pointer(-16);
            var ptr0 = passStringToWasm0(s, wasm.__wbindgen_malloc, wasm.__wbindgen_realloc);
            var len0 = WASM_VECTOR_LEN;
            var ptr1 = passArray8ToWasm0(salt, wasm.__wbindgen_malloc);
            var len1 = WASM_VECTOR_LEN;
            var ptr2 = isLikeNone(partition_id) ? 0 : passStringToWasm0(partition_id, wasm.__wbindgen_malloc, wasm.__wbindgen_realloc);
            var len2 = WASM_VECTOR_LEN;
            wasm.encryptedsearch_generateHashesForStringWithPadding(retptr, this.ptr, ptr0, len0, ptr1, len1, ptr2, len2);
            var r0 = getInt32Memory0()[retptr / 4 + 0];
            var r1 = getInt32Memory0()[retptr / 4 + 1];
            var v3 = getArrayU32FromWasm0(r0, r1).slice();
            wasm.__wbindgen_free(r0, r1 * 4);
            return v3;
        } finally {
            wasm.__wbindgen_add_to_stack_pointer(16);
        }
    }
    /**
    * Generate a version of the input string where each character has been latinized using the
    * same function as our tokenization routines.
    * @param {string} s
    * @returns {string}
    */
    static transliterateString(s) {
        try {
            const retptr = wasm.__wbindgen_add_to_stack_pointer(-16);
            var ptr0 = passStringToWasm0(s, wasm.__wbindgen_malloc, wasm.__wbindgen_realloc);
            var len0 = WASM_VECTOR_LEN;
            wasm.encryptedsearch_transliterateString(retptr, ptr0, len0);
            var r0 = getInt32Memory0()[retptr / 4 + 0];
            var r1 = getInt32Memory0()[retptr / 4 + 1];
            return getStringFromWasm0(r0, r1);
        } finally {
            wasm.__wbindgen_add_to_stack_pointer(16);
            wasm.__wbindgen_free(r0, r1);
        }
    }
}

export function __wbg_self_86b4b13392c7af56() { return handleError(function () {
    var ret = self.self;
    return addHeapObject(ret);
}, arguments) };

export function __wbindgen_object_drop_ref(arg0) {
    takeObject(arg0);
};

export function __wbg_static_accessor_MODULE_452b4680e8614c81() {
    var ret = module;
    return addHeapObject(ret);
};

export function __wbg_require_f5521a5b85ad2542(arg0, arg1, arg2) {
    var ret = getObject(arg0).require(getStringFromWasm0(arg1, arg2));
    return addHeapObject(ret);
};

export function __wbg_crypto_b8c92eaac23d0d80(arg0) {
    var ret = getObject(arg0).crypto;
    return addHeapObject(ret);
};

export function __wbg_msCrypto_9ad6677321a08dd8(arg0) {
    var ret = getObject(arg0).msCrypto;
    return addHeapObject(ret);
};

export function __wbindgen_is_undefined(arg0) {
    var ret = getObject(arg0) === undefined;
    return ret;
};

export function __wbg_getRandomValues_dd27e6b0652b3236(arg0) {
    var ret = getObject(arg0).getRandomValues;
    return addHeapObject(ret);
};

export function __wbg_newwithlength_a8d1dbcbe703a5c6(arg0) {
    var ret = new Uint8Array(arg0 >>> 0);
    return addHeapObject(ret);
};

export function __wbg_subarray_901ede8318da52a6(arg0, arg1, arg2) {
    var ret = getObject(arg0).subarray(arg1 >>> 0, arg2 >>> 0);
    return addHeapObject(ret);
};

export function __wbg_getRandomValues_e57c9b75ddead065(arg0, arg1) {
    getObject(arg0).getRandomValues(getObject(arg1));
};

export function __wbg_length_2d56cb37075fcfb1(arg0) {
    var ret = getObject(arg0).length;
    return ret;
};

export function __wbindgen_memory() {
    var ret = wasm.memory;
    return addHeapObject(ret);
};

export function __wbg_buffer_9e184d6f785de5ed(arg0) {
    var ret = getObject(arg0).buffer;
    return addHeapObject(ret);
};

export function __wbg_new_e8101319e4cf95fc(arg0) {
    var ret = new Uint8Array(getObject(arg0));
    return addHeapObject(ret);
};

export function __wbg_set_e8ae7b27314e8b98(arg0, arg1, arg2) {
    getObject(arg0).set(getObject(arg1), arg2 >>> 0);
};

export function __wbg_randomFillSync_d2ba53160aec6aba(arg0, arg1, arg2) {
    getObject(arg0).randomFillSync(getArrayU8FromWasm0(arg1, arg2));
};

export function __wbindgen_json_parse(arg0, arg1) {
    var ret = JSON.parse(getStringFromWasm0(arg0, arg1));
    return addHeapObject(ret);
};

export function __wbg_new_3e1ee746fe308c9f(arg0, arg1) {
    var ret = new Error(getStringFromWasm0(arg0, arg1));
    return addHeapObject(ret);
};

export function __wbindgen_json_serialize(arg0, arg1) {
    const obj = getObject(arg1);
    var ret = JSON.stringify(obj === undefined ? null : obj);
    var ptr0 = passStringToWasm0(ret, wasm.__wbindgen_malloc, wasm.__wbindgen_realloc);
    var len0 = WASM_VECTOR_LEN;
    getInt32Memory0()[arg0 / 4 + 1] = len0;
    getInt32Memory0()[arg0 / 4 + 0] = ptr0;
};

export function __wbindgen_throw(arg0, arg1) {
    throw new Error(getStringFromWasm0(arg0, arg1));
};

export function __wbindgen_rethrow(arg0) {
    throw takeObject(arg0);
};
